docs/plugins.md

```diff
@@ -0,0 +1,58 @@
+-   Logic to handle new transactions, validation, making changes to state, etc (for more detail look at the doc on [request handling](request_handling.md)).
+-   Storage of transactions in existing ledger or introducing new ledgers
+-   Changes to existing state or introducing new states  
+-   Defining new constants
+-   Defining any new config variables
+
+
+### Conventions that need to be followed when defining a new plugin:
+#   Each plugin is a separate python package that lives under the package defined in `PLUGIN_ROOT` in `config.py`. It is currently set to `plenum.server.plugin`
+#   The list of enabled plugins is defined in `ENABLED_PLUGINS` in `config.py`.
+#   Each plugin has unique name, this is the name in the list `ENABLED_PLUGINS` and has to match the package name of the plugin.
+#   The plugin package contains defines several items in the `__init__.py` file
+    -   It contains a ledger id for each new ledger that the plugin introduces, see `LEDGER_IDS`
+    -   For each new field that needs to be added in the `Request`, the plugin defines the name and its corresponding input validation function in a dict, see `CLIENT_REQUEST_FIELDS`
+    -   Each write or query transaction type is defined in `AcceptableWriteTypes` and `AcceptableQueryTypes` respectively
+#   Each plugin package contains a `main.py` that contains a method `integrate_plugin_in_node` to initialise and register any new ledgers, states, authenticators, 
+    request handlers, etc.
+#   If the plugin requires a new authenticator, it should be defined in `client_authn.py` in the plugin package.
+#   New storage layers should be defined in `storage.py` in the package
+#   New transactions need to be in `transactions.py` in the plugin package. To avoid conflicts in transaction type values, each plugin defines a unique prefix to be added before 
+    the value of the transaction type, see `PREFIX`.
+#   Any new config variables that have to be introduced are introduced in `get_config` in `config.py` in the plugin package.
+
+
+### Plugin setup
+When plenum initialises, initialises all enabled plugins; this is done through `setup_plugins` method defined in `plenum/__init__.py` Plenum defines 2 project level globals:
+`PLUGIN_LEDGER_IDS` which would be the set of ids of all new ledgers defined by any enabled plugin and `PLUGIN_CLIENT_REQUEST_FIELDS` would be a dictionary of all new `Request` fields
+defined by any plugin and its corresponding validator.
+`setup_plugins` iterates over each of the enabled plugin packages and updates `PLUGIN_LEDGER_IDS` and `PLUGIN_CLIENT_REQUEST_FIELDS`. 
+It then reloads some modules which defines input validation objects and schemas, this is done since set of valid ledger ids of request fields might have changed.
+
+
+### Demo plugin
+The Demo plugin serves as an example of how a plugin should be written, though it does some extra setup actions that an actual plugin will not do.
```

Comment:
Can you please explicitly mention these extra steps (ideally put them into a separate method `init_plugin_in_tests`)

Reply:
Done